dbms/src/Disks/IDisk.h

```diff
@@ -0,0 +1,207 @@
+#include <boost/noncopyable.hpp>
+
+
+namespace CurrentMetrics
+{
+extern const Metric DiskSpaceReservedForMerge;
+}
+
+namespace DB
+{
+class IDiskFile;
+using DiskFilePtr = std::shared_ptr<IDiskFile>;
+
+class DiskDirectoryIterator;
+class IDiskDirectoryIteratorImpl;
+using DiskDirectoryIteratorImplPtr = std::unique_ptr<IDiskDirectoryIteratorImpl>;
+
+class IReservation;
+using ReservationPtr = std::unique_ptr<IReservation>;
+
+class ReadBuffer;
+class WriteBuffer;
+
+/**
+ * Provide interface for reservation.
+ */
+class Space : public std::enable_shared_from_this<Space>
+{
+public:
+    virtual const String & getName() const = 0;
+
+    virtual ReservationPtr reserve(UInt64 bytes) const = 0;
+
+    virtual ~Space() = default;
+};
+
+using SpacePtr = std::shared_ptr<const Space>;
+
+/**
+ * A unit of storage persisting data and metadata.
+ * Abstract underlying storage technology.
+ * Responsible for:
+ * - file management;
+ * - space accounting and reservation.
+ */
+class IDisk : public Space
+{
+public:
+    virtual const String & getPath() const = 0;
+
+    /// Total available space on disk.
+    virtual UInt64 getTotalSpace() const = 0;
+
+    /// Space currently available on disk.
+    virtual UInt64 getAvailableSpace() const = 0;
+
+    /// Currently available (prev method) minus already reserved space.
+    virtual UInt64 getUnreservedSpace() const = 0;
+
+    /// Amount of bytes which should be kept free on this disk.
+    virtual UInt64 getKeepingFreeSpace() const { return 0; }
+
+    virtual DiskFilePtr file(const String & path) const = 0;
+};
+
+using DiskPtr = std::shared_ptr<const IDisk>;
+using Disks = std::vector<DiskPtr>;
+
+class IDiskFile : public std::enable_shared_from_this<IDiskFile>
+{
+public:
+    friend class DiskDirectoryIterator;
+
+    /// Return disk which the file belongs to.
+    const DiskPtr & disk() const { return disk_ptr; }
+
+    const String & path() const { return rel_path; }
+
+    String fullPath() const { return disk_ptr->getPath() + rel_path; }
+
+    /// Returns true if the file exists.
+    virtual bool exists() const = 0;
+
+    /// Returns true if the file is a directory.
+    virtual bool isDirectory() const = 0;
+
+    /// Creates a directory.
+    virtual void createDirectory() = 0;
+
+    /// Creates a directory and all parent directories if necessary.
+    virtual void createDirectories() = 0;
+
+    virtual void moveTo(const String & new_path) = 0;
+
+    virtual void copyTo(const String & new_path) = 0;
+
+    /// Open the file for read and returns ReadBuffer object.
+    virtual std::unique_ptr<ReadBuffer> read() const = 0;
+
+    /// Open the file for write and returns WriteBuffer object.
+    virtual std::unique_ptr<WriteBuffer> write() = 0;
+
+    virtual ~IDiskFile() = default;
+
+protected:
+    IDiskFile(const DiskPtr & disk_ptr_, const String & rel_path_) : disk_ptr(disk_ptr_), rel_path(rel_path_) {}
+
+private:
+    virtual DiskDirectoryIteratorImplPtr iterateDirectory() = 0;
```

Comment:
Why this method is here? `private` methods have to be used inside the class, but this one doesn't.

Reply:
The method was removed.